### ISSUE TITLE
key_shared subscription will get suck sometimes when add consumers

### ISSUE THREAD

**Issue by leizhiyuan** [apache/pulsar#14902]
**Describe the bug**

key_shared subscription will get suck sometimes when add consumers

when add consumers, sometimes, consumer will receive message very slow or get suck

**To Reproduce**

``` producer
package org.example;

import org.apache.pulsar.client.api.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.TimeUnit;

import static org.example.Helper.TP_TOPIC;

public class SimpleProducer {

    private static Logger logger = LoggerFactory.getLogger(SimpleProducer.class);

    public static void main(String[] args) {

        logger.info("producer start");
        try {
            execute();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void execute() throws Exception {
        PulsarClient client = Helper.buildClient();

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try {
                client.close();
            } catch (PulsarClientException e) {
                e.printStackTrace();
            }
        }));

        Producer<String> producer = client.newProducer(Schema.STRING)
                .topic(TP_TOPIC)
                .create();

        for (int i = 0; i < 10000; i++) {
            try {
                MessageId id = producer.newMessage().key(String.valueOf(i)).value(String.valueOf(i)).send();
                TimeUnit.MILLISECONDS.sleep(5);
                System.out.println(id);
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        }
    }
}

```

``` consumer
package org.example;

import static org.example.Helper.TP_TOPIC;

import java.util.Map;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

import org.apache.pulsar.client.api.Consumer;
import org.apache.pulsar.client.api.Message;
import org.apache.pulsar.client.api.MessageListener;
import org.apache.pulsar.client.api.PulsarClient;
import org.apache.pulsar.client.api.PulsarClientException;
import org.apache.pulsar.client.api.Schema;
import org.apache.pulsar.client.api.SubscriptionInitialPosition;
import org.apache.pulsar.client.api.SubscriptionType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SimpleConsumer {

    private static Logger logger = LoggerFactory.getLogger(SimpleConsumer.class);

    protected static Map<String, Consumer> consumers = new ConcurrentHashMap<>();

    static PulsarClient client = Helper.buildClient();

    public static void main(String[] args) {

        logger.info("consumer start");
        try {
            execute();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void execute() throws Exception {
        for (int i = 0; i < 50; i++) {
            createConsumer(i);
        }


        TimeUnit.SECONDS.sleep(5);

        System.out.println("update consumer num");
        for (int i = 50; i < 100; i++) {
            createConsumer(i);
        }


        TimeUnit.SECONDS.sleep(5);
        System.out.println("update consumer num");
        for (int i = 100; i < 150; i++) {
            createConsumer(i);
        }


        TimeUnit.SECONDS.sleep(1000000);


    }


    public static boolean nack() {
        if (new Random().nextInt(100) < 5) {
            return true;
        }
        return false;
    }

    private static void createConsumer(int i) throws PulsarClientException {
        Consumer consumer = client.newConsumer(Schema.BYTES)
                .topic(TP_TOPIC)
                .subscriptionName("eb-bystander")
                .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                .subscriptionType(SubscriptionType.Key_Shared)
                .messageListener(new MessageListener<byte[]>() {
                    @Override
                    public void received(Consumer<byte[]> consumer, Message<byte[]> message) {
                        System.out.println(message.getMessageId());
                        if (nack()) {
                            consumer.negativeAcknowledge(message);
                        } else {
                            try {
                                consumer.acknowledge(message);
                            } catch (PulsarClientException e) {
                                e.printStackTrace();
                            }
                        }

                    }
                })
                .subscribe();

        consumers.put(String.valueOf(i), consumer);
    }
}

```

``` helper
package org.example;

import org.apache.pulsar.client.api.ClientBuilder;
import org.apache.pulsar.client.api.PulsarClient;
import org.apache.pulsar.client.api.PulsarClientException;

public class Helper {

    public static String TP_TOPIC="persistent://public/default/zhiyuanlei";

    public static PulsarClient buildClient(){
        ClientBuilder clientBuilder = PulsarClient.builder()
                .serviceUrl("http://127.0.0.1:8080");
        PulsarClient client = null;
        try {
            client = clientBuilder.ioThreads(10).connectionsPerBroker(10).listenerThreads(10).build();
        } catch (PulsarClientException e) {
            e.printStackTrace();
        }
        return client;
    }
}

```
<img width="515" alt="image" src="https://user-images.githubusercontent.com/2684384/160319696-d6ef49c6-7aee-4c1f-a144-10bb645b9492.png">

<img width="529" alt="image" src="https://user-images.githubusercontent.com/2684384/160319995-3a698f49-ced2-4a86-b18e-bf4c83165a3f.png">

we can ensure that producer produces  without stop, but consumer seems cannot get message, 

maybe
```
 if (nack()) {
                            consumer.negativeAcknowledge(message);
                        }
``` 

this is important.

addition, in production env, we see some partition get suck, unload this topic partition can make consumer recover to normal

**Expected behavior**

consumer receives ok

**Screenshots**

[stat-internal.log](https://github.com/apache/pulsar/files/8359438/stat-internal.log)
[stat.log](https://github.com/apache/pulsar/files/8359439/stat.log)

**Desktop (please complete the following information):**

**Additional context**
client 2.7.2 / client 2.9.2
run broker in local, with master branch can reproduce.

**Comment by Jason918** [comment]
What's the broker version?

**Comment by leizhiyuan** [comment]
> What's the broker version?

2.7.2  use auto split without consistent hash

**Comment by codelipenghui** [comment]
@leizhiyuan You can check the topic stats if you can reproduce the problem, and share the topic stats/internal-stats here.

**Comment by leizhiyuan** [comment]
> @leizhiyuan You can check the topic stats if you can reproduce the problem, and share the topic stats/internal-stats here.

ok， next time I will keep it

**Comment by leizhiyuan** [comment]
[stats.log](https://github.com/apache/pulsar/files/8769552/stats.log)
[stats_internal.log](https://github.com/apache/pulsar/files/8769553/stats_internal.log)

there are two subscriptions , one is key shared, the other is shared, shared is ok

**Comment by lhotari** [comment]
Fixed by [PIP-379](https://github.com/apache/pulsar/blob/master/pip/pip-379.md), #23352. The fix will be available in Pulsar 4.0.0+.
